opamp-client/src/main/java/io/opentelemetry/opamp/client/internal/OpampClient.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.opamp.client.internal;
+
+import io.opentelemetry.opamp.client.internal.response.MessageData;
+import opamp.proto.Opamp;
+
+public interface OpampClient {
+
+  /**
+   * Starts the client and begin attempts to connect to the Server. Once connection is established
+   * the client will attempt to maintain it by reconnecting if the connection is lost. All failed
+   * connection attempts will be reported via {@link Callback#onConnectFailed(OpampClient,
+   * Throwable)} callback.
+   *
+   * <p>This method does not wait until the connection to the Server is established and will likely
+   * return before the connection attempts are even made.
+   *
+   * <p>This method may be called only once.
+   *
+   * @param callback The Callback to which the Client will notify about any Server requests and
+   *     responses.
+   */
+  void start(Callback callback);
+
+  /**
+   * Stops the client. May be called only after {@link #start(OpampClient.Callback)}. May be called
+   * only once. After this call returns successfully it is guaranteed that no callbacks will be
+   * called. Once stopped, the client cannot be started again.
+   */
+  void stop();
+
+  /**
+   * Sets the current remote config status which will be sent in the next agent to server request.
+   *
+   * @param remoteConfigStatus The new remote config status.
+   */
+  void setRemoteConfigStatus(Opamp.RemoteConfigStatus remoteConfigStatus);
+
+  /**
+   * Sets the current effective config which will be sent in the next agent to server request.
+   *
+   * @param effectiveConfig The new effective config.
+   */
+  void setEffectiveConfig(Opamp.EffectiveConfig effectiveConfig);
```

Comment:
In opamp-go effective config is provided via a callback. I know it is not consistent with the rest, but there is a reason we went that way, see comments here https://github.com/open-telemetry/opamp-go/blob/b33ab76c401d0f5c3f179c5905409c1502176f97/client/internal/clientstate.go#L26

Happy to expand more if the comments are not clear.

Reply:
I must admit I thought it was strange at first because it wasn't consistent and I also missed that comment, so thank you for pointing it out 🙏 

Now that I'm aware of this, I'm curious if there could be a way to address this issue while keeping some consistency with the other values at the same time 🤔 I'll have a look, and if I can't find a way I'll add the callback option 👍

Reply:
Excellent. I am happy to hear any thoughts you have about how we can make this better. I will gladly take it back to opamp-go.

Reply:
I believe I found a way to address this, I also like the overall impl better thanks to the changes I made for it to work so I'm glad we had this discussion.

The approach combines a [Supplier](https://docs.oracle.com/javase/8/docs/api/index.html?java/util/function/Supplier.html) and the observer pattern to create a [State](https://github.com/elastic/opamp-java/blob/main/client/src/main/java/co/elastic/opamp/client/state/State.java#L31) type which will be used for [all the AgentToServer fields](https://github.com/elastic/opamp-java/blob/main/client/src/main/java/co/elastic/opamp/client/internal/state/OpampClientState.java#L24).

For most of the fields, the State implementation they will use is [InMemoryState](https://github.com/elastic/opamp-java/blob/main/client/src/main/java/co/elastic/opamp/client/internal/state/InMemoryState.java#L23), which will store its value in memory, but for the effective config state, users will have to provide their own implementation when [building the client](https://github.com/elastic/opamp-java/blob/main/client/src/main/java/co/elastic/opamp/client/OpampClientBuilder.java#L155).

When it's built, the client [observes changes from these states](https://github.com/elastic/opamp-java/blob/main/client/src/main/java/co/elastic/opamp/client/internal/OpampClientImpl.java#L211) to know when to add them into the [next request](https://github.com/elastic/opamp-java/blob/main/client/src/main/java/co/elastic/opamp/client/internal/OpampClientImpl.java#L198). That way users can tell from within their own state implementation when should the client query it and send the effective config field.

I created this [dummy implementation](https://github.com/elastic/opamp-java/blob/main/sample-app/src/main/java/co/elastic/opamp/sample/MyEffectiveConfigState.java) which is used in [this example](https://github.com/elastic/opamp-java/blob/main/sample-app/src/main/java/co/elastic/opamp/sample/MainWebsocket.java#L21) to show how it could be used.

The consequence of this is that the `OpampClient` interface won't have any setter/updater method related to the effective config because that state isn't handled by the client, so the user should keep their provided implementation somewhere handy for them to interact with it and update it/request for it to be sent in the next request. For the values that we want to store in memory though, the client keeps the setter methods as a convenient way of setting new values into the InMemoryState implementations, like [this one](https://github.com/elastic/opamp-java/blob/main/client/src/main/java/co/elastic/opamp/client/internal/OpampClientImpl.java#L120) for example. I'm not sure if we should keep these convenience methods though because probably some users will get confused about why not all fields have setters in the client, but the alternative is to make users to provide all the state implementations and keep them all somewhere handy so that they can change them at some point, so I guess the convenience methods can be very handy to make things easier in general.

I'm open to discussing more details about it if needed and also looking forward to any feedback! Cheers.